Add option to display timestamps in the local system timezone

jparismorgan opened the pull request:
### What
Adds the option to display timestamps in the local system timezone, addressing https://github.com/rerun-io/rerun/issues/1714.

### Checklist
* [x] I have read and agree to [Contributor Guide](https://github.com/rerun-io/rerun/blob/main/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/rerun-io/rerun/blob/main/CODE_OF_CONDUCT.md)
* [ ] I've included a screenshot or gif (if applicable)
* [ ] I have tested [demo.rerun.io](https://demo.rerun.io/pr/{{ pr.number }}) (if applicable)

- [PR Build Summary](https://build.rerun.io/pr/{{ pr.number }})
- [Docs preview](https://rerun.io/preview/{{ pr.commit }}/docs) <!--DOCS-PREVIEW-->
- [Examples preview](https://rerun.io/preview/{{ pr.commit }}/examples) <!--EXAMPLES-PREVIEW-->
- [Recent benchmark results](https://ref.rerun.io/dev/bench/)
- [Wasm size tracking](https://ref.rerun.io/dev/sizes/)

jparismorgan commented:
Hi there! I've quite liked watching Rerun be developed over the past year, so I wanted to contribute a bit. This is also my first Rust I've written, it's quite interesting how similar yet different it is to C++. 

Anyways, I'm opening up this PR as an early draft for feedback - does it line up with what you were expecting? A few specific questions / thoughts:
1. The pattern of passing `show_timestamps_in_local_timezone` is somewhat cumbersome in that everyone must now pass the flag. Is there another suggested way to do this?
2. I hard-coded `show_timestamps_in_local_timezone` in several locations in `arrow_store/`. I believe this is correct because this is the code to serialize data, but perhaps I missed something there?
3. `Cargo.lock` had a new dependency added, I believe because I added `local-offset` here: `time = { workspace = true, features = ["formatting", "macros", "local-offset"] }`. Can test to see if that's the case if this type of change is unexpected.
4. I have a failing unit test locally which looks like it's in `crates/re_sdk/src/recording_stream.rs`. Still debugging if it's related:
```
(rerun-jparismorgan-scripts) ~/repo/rerun-jparismorgan cargo test --all-targets --all-features -p re_sdk --lib                             local-time-option 
   Compiling re_sdk v0.9.0-alpha.4 (/Users/parismorgan/repo/rerun-jparismorgan/crates/re_sdk)
    Finished test [optimized + debuginfo] target(s) in 1.30s
     Running unittests src/lib.rs (target/debug/deps/re_sdk-2385fcf4b104c75d)

running 6 tests
test recording_stream::tests::impl_send_sync ... ok
test recording_stream::tests::disabled ... ok
thread panicked while processing panic. aborting.
test recording_stream::tests::always_flush ... ok
test recording_stream::tests::flush_hierarchy ... ok
test recording_stream::tests::never_flush ... ok
error: test failed, to rerun pass `-p re_sdk --lib`

Caused by:
  process didn't exit successfully: `/Users/parismorgan/repo/rerun-jparismorgan/target/debug/deps/re_sdk-2385fcf4b104c75d` (signal: 5, SIGTRAP: trace/breakpoint trap)
```
5. Finally, there are quite a few TODO's left to do - would be curious if you know the answer to any. I was mainly working quickly and will come back to them once the general structure looks good.

Thank you!

PS. Here's a screenshot of work so far - you can see we still have a UTC date next to `rerun_example_detect_and_track_objects`, but the `Recording started` date is correct: 

<img width="1512" alt="Screenshot 2023-09-29 at 12 44 07 AM" src="https://github.com/rerun-io/rerun/assets/1396242/2547c217-53bc-4314-bf1e-8b4458c2af56">

And here's the option menu

<img width="583" alt="Screenshot 2023-09-29 at 12 58 38 AM" src="https://github.com/rerun-io/rerun/assets/1396242/1ef64498-ef99-4e81-9bf6-ad8ed9a2a3d1">

emilk commented:
Hi, and thank you so much for this PR!

I strongly suggest switching out the boolean for an enum:

```rust
#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
enum TimeZone {
    Local,
    Utc,
}
```

This will make the code more readable, but also allow us to easily add custom timezones in the future, if we so desire.

Piping this explicitly is 

The store mostly uses time formatting for trace logging and error messages, and for that UTC is fine. You can create a helper `format_utc()` so you don't need to pass in `TimeZone::Utc` in all these places.

emilk commented on crates/re_log_types/src/time.rs:
```diff
@@ -63,19 +63,23 @@ impl Time {
         let nanos_since_epoch = self.nanos_since_epoch();
 
         if let Some(datetime) = self.to_datetime() {
+            let local_offset = UtcOffset::current_local_offset();
+            let will_show_timestamps_in_local_timezone =
+                show_timestamps_in_local_timezone && local_offset.is_ok();
```
```suggestion
            let local_offset = show_timestamps_in_local_timezone.then(|| UtcOffset::current_local_offset());
```

jparismorgan replied:
Thanks! With `let local_offset = (time_zone_for_timestamps == TimeZone::Local).then(|| UtcOffset::current_local_offset());` I get:
```
error: redundant closure
   --> crates/re_log_types/src/time.rs:154:87
    |
154 |                 let local_offset = (time_zone_for_timestamps == TimeZone::Local).then(|| UtcOffset::current_local_offset());
    |                                                                                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: replace the closure with the function itself: `UtcOffset::current_local_offset`
```
Changing to `let local_offset = (time_zone_for_timestamps == TimeZone::Local).then(UtcOffset::current_local_offset);` fixes this. Any concerns?

emilk commented on crates/re_log_types/src/time.rs:
```diff
@@ -85,7 +89,17 @@ impl Time {
             };
-            datetime.format(&parsed_format).unwrap()
+
+            // Return in the local timezone.
+            if will_show_timestamps_in_local_timezone {
+                let local_datetime = datetime.to_offset(local_offset.unwrap());
+                return local_datetime.format(&parsed_format).unwrap();
+            }
+
+            // Return in UTC.
+            let mut formatted_datetime = datetime.format(&parsed_format).unwrap();
+            formatted_datetime.push('Z');
+            formatted_datetime
```
```suggestion
            if let Some(local_offset) = local_offset {
                // Return in the local timezone.
                let local_datetime = datetime.to_offset(local_offset);
                local_datetime.format(&parsed_format).unwrap()
            } else {
                // Return in UTC.
                format!("{}Z", datetime.format(&parsed_format).unwrap())
            }
```

emilk replied:
same in the other formatting place

jparismorgan replied:
Thanks! I think we also need an `Ok(` in there, i.e. `if let Some(Ok(local_offset)) = local_offset {`, correct? Because `Some(` will just unwrap the `Option` but not check the `Result`?

<img width="976" alt="Screenshot 2023-09-29 at 2 20 08 PM" src="https://github.com/rerun-io/rerun/assets/1396242/ec234feb-9fa8-46fa-a102-56f96aeca1af">

emilk commented on crates/re_time_panel/src/paint_ticks.rs:
```diff
@@ -68,7 +76,8 @@ fn paint_time_range_ticks(
                 next_grid_tick_magnitude_ns,
-                |ns| Time::from_ns_since_epoch(ns).format_time_compact(),
+                // TODO(paris): Fix 'closures can only be coerced to `fn` types if they do not capture any variables' error.
+                |ns| Time::from_ns_since_epoch(ns).format_time_compact(false),
```
You can change `paint_ticks` from taking a `format_tick: fn(i64) -> String` into taking a `format_tick: impl Fn(i64) -> String`

jparismorgan replied:
Great thanks, updated.

Wumpf commented on crates/re_arrow_store/src/store_format.rs:
```diff
@@ -123,8 +126,9 @@ impl std::fmt::Display for IndexedBucket {
 
         let time_range = {
             let time_range = &self.inner.read().time_range;
+            // TODO(paris): Understand if this is correct or if we should use show_timestamps_in_local_timezone. Same with other store_* files.
```
For displaying a raw bucket directly via the display trait this is fine.
Generally I think UTC is a good default for the Display trait where we naturally can't take user preferences into account.
In the long run more and more ui should be more specific about its needs calling dedicated methods which is opposed to "dump this data into a string for human consumption". 
Implies ofc that `std::fmt::Display` soon won't do the trick in a lot of situations.

jparismorgan replied:
Great thanks, leaving in UTC.

Wumpf commented on crates/re_build_tools/src/lib.rs:
```diff
@@ -85,6 +85,7 @@ pub fn export_env_vars() {
         );
     }
 
+    // TODO(paris): Determine if this needs to be updated if we're displaying in the local timezone.
```
This is used for etching build time into the build itself and should stay in UTC.

jparismorgan replied:
Sounds good, thanks.

emilk commented on crates/re_viewer/src/ui/rerun_menu.rs:
```diff
@@ -164,6 +164,19 @@ impl App {
             ui.close_menu();
         }
 
+        if self
+            .re_ui
+            .checkbox(
+                ui,
+                &mut self.state.app_options.show_timestamps_in_local_timezone,
+                "Show timestamps in local timezone",
+            )
+            .on_hover_text("Convert timestamps from UTC to the local timezone")
+            .clicked()
+        {
+            ui.close_menu();
+        }
```
I think this would be nicer as two radio buttons:

```
( ) UTC    (·) Local timezone
```

jparismorgan replied:
Done!

<img width="274" alt="Screenshot 2023-09-29 at 1 10 54 PM" src="https://github.com/rerun-io/rerun/assets/1396242/50da9f8a-07e8-4af6-9f76-02772b2e8e81">

Wumpf commented on crates/re_log_types/src/data_row.rs:
```diff
@@ -546,7 +546,15 @@ impl std::fmt::Display for DataRow {
         writeln!(f, "Row #{} @ '{}'", self.row_id, self.entity_path)?;
         for (timeline, time) in &self.timepoint {
-            writeln!(f, "- {}: {}", timeline.name(), timeline.typ().format(*time))?;
+            // TODO(paris): Figure how to pass show_timestamps_in_local_timezone in to fmt(). We get:
```
as with `IndexedBucket` I don't think we need to care right now about changing the display for `DataRow`. For any more sophisticated ui we have a look at what's inside a row first. Making UTC a good default again.

jparismorgan replied:
Great thanks, left it in UTC.

Wumpf commented on crates/re_time_panel/src/paint_ticks.rs:
```diff
@@ -68,7 +76,8 @@ fn paint_time_range_ticks(
                 time_range, // ns
                 next_grid_tick_magnitude_ns,
-                |ns| Time::from_ns_since_epoch(ns).format_time_compact(),
+                // TODO(paris): Fix 'closures can only be coerced to `fn` types if they do not capture any variables' error.
+                |ns| Time::from_ns_since_epoch(ns).format_time_compact(false),
```
the problem is with the signature of `paint_ticks`. Right now it takes the equivalent of a C function pointer:
`format_tick: fn(i64) -> String,`

Change it to `format_tick: impl Fn(i64) -> String` essentially making it a template method so it can take arbitrary objects that fullfil the `Fn(i64) -> String` contract (like passing a functor or a lambda in C++)

jparismorgan replied:
I see great, that works, thank you.

Wumpf reviewed (commented):
First of all thank you so much for contributing and tackling this feature! And welcome to Rust 😄 .

> 1) I went with the pattern of passing show_timestamps_in_local_timezone into format(), which seems easiest to keep the function stateless. But curious if there's another pattern to do this I missed in the codebase?

I think that's the correct way to got about it here, but instead of making it a an enum (something like `enum TimeFormat`?) would make things more readable and extensible :)

> 2) I hard-coded show_timestamps_in_local_timezone = false in several locations in arrow_store/. I believe this is correct because this is the code to serialize data, but perhaps I missed something there?

Didn't go through all the callsites yet, but it looks like this affects mostly debug printing, tests and examples internal to `arrow_store`. So yes, for those the UTC behavior seems best 👍 


> 3) Cargo.lock had a new dependency added, I believe because I added local-offset here: time = { workspace = true, features = ["formatting", "macros", "local-offset"] }.

Digged after this a bit and can confirm. The `local-offset` feature drags this dependency in on linux https://github.com/time-rs/time/blob/main/time/Cargo.toml#L62. But it seems to be super small, so that's alright.


> 4) I have a failing unit test locally which looks like it's in crates/re_sdk/src/recording_stream.rs.

I actually have the same failure on main running locally Oo. Let me check what's up there...

> 5) Finally, there are quite a few TODO's left to do 

I left some comments 👍 



General direction works I think. Passing the setting through every corner is annoying on one hand but more flexible in the long run. Relatedly, I'm not sure yet where we want to store the setting in the long run and how granular this kind of thing should be configurable but for the moment it's definitely the right place.

As you already know, if you need quick feedback or answers to question where the Github format doesn't do you can always post on Discord as well :)

Wumpf commented:
looks like @emilk wrote much of the same while I was busy 😅

jparismorgan commented:
Thanks @Wumpf and @emilk! I've changed to an enum and cleaned up the code a bit. Still have a few TODO's in there if you'd like to take another pass, they are mainly around Rust & Rerun patterns.

emilk commented on crates/re_arrow_store/src/store_read.rs:
```diff
@@ -750,7 +751,7 @@ impl IndexedBucket {
                     ?components,
                     timeline = %self.timeline.name(),
-                    time = self.timeline.typ().format(time),
+                    time = self.timeline.typ().format(time, TimeZone::Utc),
```
This is such a common pattern that I would suggest we add `format_utc`:

```suggestion
                    time = self.timeline.typ().format_utc(time),
```

jparismorgan replied:
Sure, done!

Wumpf commented on crates/re_log_types/src/lib.rs:
```diff
@@ -52,7 +52,8 @@ pub use self::data_table::{
 pub use self::index::*;
 pub use self::path::*;
-pub use self::time::{Duration, Time};
+// TODO(paris): TimeZone is also used by Timeline etc. in time_point - should I put it under a different parent / a new top level parent?
```
that's fine, we can leave it there for the time being (pun not intended)

jparismorgan replied:
Great, thanks.

Wumpf commented on crates/re_log_types/src/time.rs:
```diff
@@ -1,5 +1,16 @@
-use time::{OffsetDateTime, UtcOffset};
+use time::{format_description::FormatItem, OffsetDateTime, UtcOffset};
+
+// TODO(paris): Added `serde::Deserialize, serde::Serialize` to fix error in `crates/re_viewer_context/src/app_options.rs`:
```
makes sense, that is indeed needed for having it in the app options sine app options is serialized with serde, so everything it contains needs to be serializable as well. Your solution is correct

jparismorgan replied:
Sounds good, thanks.

Wumpf commented on crates/re_log_types/src/time.rs:
```diff
@@ -62,15 +73,30 @@ impl Time {
     }
 
+    // TODO(paris): Is there a pattern in Rust where you put free functions in an anonymous namespace versus in the struct?
+    // TODO(paris): Should we #[inline] this? I'm guessing the suggestion is to benchmark with and without, though this perhaps seems a bit big for inlining.
```
Formatting is unlikely to be time critical for us, so I wouldn't put #inline without reason :)

jparismorgan replied:
Sounds good, left as-is.

Wumpf commented on crates/re_log_types/src/time.rs:
```diff
@@ -62,15 +73,30 @@ impl Time {
+        parsed_format: &Vec<FormatItem<'_>>,
+        time_zone_for_timestamps: TimeZone,
+    ) -> String {
+        let local_offset =
+            (time_zone_for_timestamps == TimeZone::Local).then(UtcOffset::current_local_offset);
```
a better pattern here would be to do a `match` over the timezone settings. That way it stops compiling when somebody adds a new enum value but doesn't handle it here :)

```rust
match time_zone_for_timestamp {
    TimeZone::Local => { ... }
     TimeZone::Utc => { ... }
}
```

jparismorgan replied:
Yep good call, updated. Leads to some repeat of code now, but seems worth it to avoid future issues.

Wumpf reviewed (commented):
(brief scroll over new `TODO` notes / ones I might have missed :))

jparismorgan commented on crates/re_log_types/src/time.rs:
```diff
@@ -175,6 +178,24 @@ impl Time {
         }
     }
 
+    // TODO(paris): I left the return type as Option<String> to match the existing calling code in recordings_panel.rs. Is this okay, or a bad pattern?
```
Sounds good, left code as-is.

jparismorgan commented on crates/re_log_types/src/time.rs:
```diff
@@ -62,15 +73,30 @@ impl Time {
         self.nanos_since_epoch() % (24 * 60 * 60 * 1_000_000_000) == 0
     }
 
+    // TODO(paris): Is there a pattern in Rust where you put free functions in an anonymous namespace versus in the struct?
```
Sounds good, left as-is.

jparismorgan commented:
Okay so removed the last of my TODO's and ready for another round of review! 

https://demo.rerun.io/pr/3530 is not yet live - I believe because some of the workflows require approval from a maintainer to start?

<img width="815" alt="Screenshot 2023-10-02 at 12 23 59 PM" src="https://github.com/rerun-io/rerun/assets/1396242/3bc00f45-6f8f-42e3-ab0c-80f2a83ee089">

Wumpf commented:
> https://demo.rerun.io/pr/3530 is not yet live

web demo upload is sadly one of the things we can't do yet for PRs from fork since it queries a bunch of tokens that we don't want to leak to a github action that we don't own 🤔 
you can ignore that part of the pr template

emilk commented on crates/re_arrow_store/src/store_format.rs:
```diff
@@ -124,7 +128,7 @@ impl std::fmt::Display for IndexedBucket {
             let time_range = &self.inner.read().time_range;
             if time_range.min.as_i64() != i64::MAX && time_range.max.as_i64() != i64::MIN {
-                self.timeline.format_time_range(time_range)
+                self.timeline.format_time_range(time_range, TimeZone::Utc)
```
could be `format_range_utc`

jparismorgan replied:
Sure, thanks, done.

Wumpf reviewed (changes requested):
aaaalmost good to go. Looking really good now :)

Wumpf commented:
Tested it locally on the web, works fine on Firefox & Chrome on my Mac :)

jparismorgan commented on crates/re_log_types/src/time.rs:
```diff
@@ -62,20 +69,44 @@ impl Time {
+                } else {
+                    // Return in UTC.
+                    format!("{}Z", datetime.format(&parsed_format).unwrap())
+                }
```
Oops yes, good catch, thank you, missed removing that check. Updatd.

jparismorgan commented on crates/re_log_types/src/time.rs:
```diff
@@ -146,6 +179,23 @@ impl Time {
+                time_zone_for_timestamps,
+            ));
+        }
+        None
```
Nice, thanks, updated.

emilk replied:
Or just use `?`

Wumpf reviewed (approved):
Thank you again!

All good from my side. I merged in main because we have some ci issues here that I thought we fixed there but turns out they still weren't configured correctly. @jprochazk is taking care of it right now :)
Will merge once we achieved green ci!

Wumpf commented:
oh, looks like your new test actually fails on CI right now:
```
'time::tests::test_formatting_local_time_zone_no_z' panicked at 'assertion failed: !&datetime.format(TimeZone::Local).contains(\'Z\')', crates/re_log_types/src/time.rs:388:9
```
https://github.com/rerun-io/rerun/actions/runs/6382347518/job/17320741545?pr=3530

jparismorgan commented:
> oh, looks like your new test actually fails on CI right now:
> 
> ```
> 'time::tests::test_formatting_local_time_zone_no_z' panicked at 'assertion failed: !&datetime.format(TimeZone::Local).contains(\'Z\')', crates/re_log_types/src/time.rs:388:9
> ```
> 
> https://github.com/rerun-io/rerun/actions/runs/6382347518/job/17320741545?pr=3530
> 
> I suspect it failed to determine the time offset on CI?

Thanks, yep looks like it. I pushed a commit which added some logging, but I'm not seeing  it printed out (https://github.com/rerun-io/rerun/actions/runs/6384854632/job/17329151635). So seems I need to update something about the CI also to get it to print. Looking more into that...

jprochazk commented:
> So seems I need to update something about the CI also to get it to print. Looking more into that...

The test needs `re_log::setup_native_logging` at the start. That might not be enough, I don't remember what the default log levels are. I think you may have more success replacing all the `re_log` calls with plain [`println!`](https://doc.rust-lang.org/std/macro.println.html)

Wumpf commented:
If we can't get this to work on CI I'm very much open to disabling the test since clearly it worked on a variety of systems. But we should probably make the "local time" option greyed out in the ui if we know it's not available, we can do so in a follow-up
Would be ofc nice to find out what about the CI config causes it to fail to get the UTC offset, but this might get tricky

jparismorgan commented:
Makes sense, thanks. I removed the test (that's what you meant by disable, right?) and in a follow-up will disable the radio boxes if we fail to get the local timezone.

Wumpf commented:
commented out or something would have been fine, but this will do. Waiting for final final ci run and then we can ship this as part of 0.9 later this week 😄
